### PR TITLE
Upgrade ESLint and eslint-config-google to latest

### PR DESCRIPTION
**Pull request by paulirish**
Gives us some nice benefits including newer rules that have been shipped recently. One-such is `prefer-const`.

[The total diff in eslint-config-google is here.](https://github.com/google/eslint-config-google/compare/7e75458986cf14982e2e3bd0e964030a8575ecb3...0f3fe777445d3c0cab21fd7daa4b5af6760b7880#diff-168726dbe96b3ce427e7fedce31bb0bc)

Another nice new feature is the `codeframe` formatter which illuminates the problem area quite clearly:
image

I've turned this formatter on for our `yarn run lint` command.  

-------------------------

Side note: I really prefer running our `npm scripts` via `yarn` rather than `npm`. Because text:
image

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -24,11 +24,17 @@
...
     }],
+    "brace-style": [2, "1tbs", {
+      "allowSingleLine": true
+    }],
     "quotes": [2, "single"],
     "require-jsdoc": 0,
     "valid-jsdoc": 0,
     "comma-dangle": 0,
     "prefer-const": 2,
+    // browser/node support for rest params isn't ready yet
```
Comment (ebidel) [inline .eslintrc.js]: I'd make the comment more explicit. It's node 4 that's the issue, right?

File: package.json [hunk 2]
```diff
@@ -31,17 +31,16 @@
     "babel-core": "^6.16.0",
     "babel-plugin-transform-es2015-destructuring": "^6.9.0",
     "coveralls": "^2.11.9",
-    "eslint": "^2.4.0",
-    "eslint-config-google": "^0.4.0",
-    "google-closure-compiler": "^20160517.0.0",
+    "eslint": "^3.12.0",
+    "eslint-config-google": "^0.7.1",
+    "google-closure-compiler": "^20160911.0.0",
```
Comment (ebidel) [inline package.json]: any reason not to use the latest from 12/1?
https://github.com/google/closure-compiler/wiki/Releases#december-1-2016-v20161201

File: .eslintrc.js (hunk 1 above)
Comment (paulirish) [inline .eslintrc.js]: sg. 
its node 5 without a flag that's the problem for us.

we haven't officially retired support but we could soon.

**Comment by paulirish** [comment]
ptal

**Comment by brendankenny** [comment]
why wasn't prefer-const completely working before? Bug in eslint?

File: .eslintrc.js [hunk 3]
```diff
@@ -24,11 +24,17 @@
       "argsIgnorePattern": "(^reject$|^_$)",
       "varsIgnorePattern": "(^_$)"
     }],
+    "brace-style": [2, "1tbs", {
+      "allowSingleLine": true
```
Comment (brendankenny) [inline .eslintrc.js]: did you want
```js
if (foo) bar();
```
or
```js
if (foo) { bar(); }
```

Comment (paulirish) [inline .eslintrc.js]: aye. i was going for `if (foo) bar();` which is handled under `curly` instead. And the style we agreed on last call isn't an option: http://eslint.org/docs/rules/curly

so i'll revert this bit.

**Comment by paulirish** [comment]
> why wasn't prefer-const completely working before? Bug in eslint?

AFAIK, it was detecting when you were using let and should be using const. but it didn't apply the same check to `var`. And now it does. bug in eslint, basically yeah.

**Pull request merged**